core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -313,6 +303,33 @@ final class ShuffleBlockFetcherIterator(
   }
 }
 
+// Helper class that ensures a ManagedBuffer is released on InputStream.close()
+private class WrappedInputStream(
+              delegate: InputStream, buf: ManagedBuffer, var currentResult: FetchResult)
+  extends InputStream {
+
+  override def read(): Int = delegate.read()
+
+  override def close(): Unit = {
+    delegate.close()
```

Comment:
Our overridden close method needs to be idempotent (since non-idempotent close() methods have been a frequent source of bugs).  I'd add an `if(!closed)` to guard this so that `release()` isn't called twice on a double-close.